### PR TITLE
Add `fix_aspect_ratio` fn

### PR DESCRIPTION
**Pull request by derekdreery**
This is a function that I would find useful. It gives you a `Rect` to draw on that has a specific aspect ratio, is centred in the containing `Rect`, and is the maximum size it can be. Name is bikesheddable if you like the function.

### CODE REVIEW

**Comment by derekdreery** [comment]
It's quite hard to explain in text what this function does, but it's very easy in pictures. I made it because to draw a pie chart I need, given any containing rect, a square that looks sensible and fits inside the rect.

**Comment by derekdreery** [comment]
Ok so I think this PR is better now. It doesn't panic and it has a better name. I've added tests to demonstrate how it performs with edge cases.

File: src/rect.rs [hunk 1]
```diff
@@ -615,4 +656,38 @@ mod tests {
         );
     }
     */
+
+    #[test]
+    fn fix_aspect_ratio() {
+        use std::f64;
```
Comment (xStrom) [inline src/rect.rs:662]: The Travis CI here uses a rather old rust compiler version. I think a better approach would be to improve the CI situation. I'll see if I can cook something up.

Comment (derekdreery) [reply src/rect.rs:662]: I'm happy to do this as an alternative, what are others' thoughts?

Comment (cmyr) [reply src/rect.rs:662]: @raphlinus was it an intentional decision to support older rust compilers? I seem to recall some conversation to that effect but I might be confused.

Comment (raphlinus) [reply src/rect.rs:662]: I think not so much an intentional decision to support older compilers, but so we know when to explicitly raise the MSRV. I'm fine with requiring latest stable actually.

At some point in the future (when people want to distribute druid apps on Debian, for example), we might want to have a more conservative MSRV policy.

File: src/rect.rs [hunk 2]
```diff
@@ -420,6 +420,47 @@ impl Rect {
...
+    /// let inner = outer.fix_aspect_ratio(1.0);
+    /// // The new `Rect` is a square centered at the center of `outer`.
+    /// assert_eq!(inner, Rect::new(0.0, 5.0, 10.0, 15.0));
+    /// ```
+    ///
+    pub fn fix_aspect_ratio(&self, aspect_ratio: f64) -> Rect {
+        let (width, height) = (self.width(), self.height());
+        let self_aspect = height / width;
+
+        if self_aspect == aspect_ratio {
```
Comment (cmyr) [inline src/rect.rs:445]: This seems like a place where floating point error could come up, although the downside is negligible.

File: src/rect.rs [hunk 3]
```diff
@@ -420,6 +420,47 @@ impl Rect {
     pub fn to_rounded_rect(self, radius: f64) -> RoundedRect {
         RoundedRect::from_rect(self, radius)
     }
+
+    /// Returns the largest possible `Rect` that is fully contained in this rect with the fixed
+    /// aspect ratio given by `aspect_ratio`.
```
Comment (cmyr) [inline src/rect.rs:425]:
```suggestion
```suggestion
    /// Returns the largest possible `Rect` that is fully contained in `self` with the given `aspect_ratio`.
    ///
    /// The aspect ratio is specified fractionally, as `height / width`.
```
    /// aspect ratio given by `aspect_ratio`.
```

File: src/rect.rs [hunk 4]
```diff
@@ -420,6 +420,47 @@ impl Rect {
...
+    ///
+    /// ```
+    /// # use kurbo::Rect;
+    /// let outer = Rect::new(0.0, 0.0, 10.0, 20.0);
+    /// let inner = outer.fix_aspect_ratio(1.0);
+    /// // The new `Rect` is a square centered at the center of `outer`.
+    /// assert_eq!(inner, Rect::new(0.0, 5.0, 10.0, 15.0));
+    /// ```
+    ///
+    pub fn fix_aspect_ratio(&self, aspect_ratio: f64) -> Rect {
```
Comment (cmyr) [inline src/rect.rs:441]: The only real Q for me about this PR is about naming.

`constrain_aspect_ratio`? `contained_rect_with_aspect_ratio`? `with_fixed_aspect_ratio`?

Comment (derekdreery) [reply src/rect.rs:441]: i think probably a longer name that fully describes the function would be better - so `contained_rect_with_aspect_ratio`, what do others think?

Comment (cmyr) [reply src/rect.rs:441]: I agree that clarity is the thing to optimize for here, that name is meh but also it's fine, happy to get this merged with that change (and optionally something along the lines of the doc suggestion above) and then we can get it out and into the next piet & druid releases.

File: src/rect.rs (hunk 2 above)
Comment (derekdreery) [reply src/rect.rs:445]: Is there a `fuzzy_eq` or similar?

File: src/rect.rs (hunk 4 above)
Comment (derekdreery) [reply src/rect.rs:441]: It can always be changed (or removed) before 1.0.

File: src/rect.rs (hunk 2 above)
Comment (derekdreery) [reply src/rect.rs:445]: I've done `(x - y).abs() < 1e-6` for now - is there a standard negligible error you use?

Comment (cmyr) [reply src/rect.rs:445]: no, you'd have to just pick a threshold and check against that. We do really want to have some sort of comprehensive fuzzy_eq implementation but we haven't done it yet.

**Comment by derekdreery** [comment]
I've changed the name, and also rebased + squashed.

**Comment by cmyr** [comment]
hmm I screwed something up with that suggestion 🤔

**Comment by derekdreery** [comment]
Right I think it's good now. :)

**Comment by derekdreery** [comment]
Ohhh ok clippy doesn't want == with float - I will use 1e-9 and make a note that the parameter hasn't been tuned.

File: src/rect.rs (hunk 2 above)
Comment (xStrom) [reply src/rect.rs:447]: I've found [float-cmp](https://crates.io/crates/float-cmp) to be pretty good, but admittedly my floating point understanding is superficial.

**Comment by cmyr** [comment]
sure, you can also just do a `#[allow(_)]` on the method.

**Comment by derekdreery** [comment]
I put a pathological example in and apparently the output has changed in master. I've just removed it as it was a 0/infinity situation, where any answer is meaningless. The fact that it is now `NaN` is probably better than before.

**Comment by cmyr** [comment]
Can you clarify what you mean about the output having changed?

**Comment by cmyr** [comment]
okay let's get this in, would like to get a new kurbo out for druid 0.6. Thanks again @derekdreery!

**Comment by derekdreery** [comment]
In the case where the width was 0 and the aspect ratio was infinity, some
of the results became NaN rather than 0. It was a pathological case though
so any results would be meaningless
